### ISSUE TITLE
const-eval allocation missing in dead_alloc_map

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#55287]
Spawned off of @oli-obk 's [example](https://github.com/rust-lang/rust/issues/55223#issuecomment-431811751) on #55223:

[play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015&gist=b23ea8c8d50f4a31a52780e27ef55fa5)

```rust
#![feature(const_let)]

union Foo<'a> {
    y: &'a (),
    long_live_the_unit: &'static (),
}

const FOO: &() = {
    let y = ();
    unsafe { Foo { y: &y }.long_live_the_unit }
};

fn main() {
}
```

**Comment by pnkfelix** [comment]
@oli-obk to be clear: we don't *need* to use `feature(const_let)` to reproduce the problem, right? I.e. hypothetically examples exist that don't use that feature?

In particular I know this problem is in some way related to #55288, which is *definitely* a true regression of stable code.

**Comment by oli-obk** [comment]
Right, this here is not really a regression. It is just an ICE that should be an error and it is fixed by https://github.com/rust-lang/rust/pull/55262

We can probably do this without const_let, but I couldn't come up with an example.

**Comment by kornelski** [comment]
A user is [reporting](https://users.rust-lang.org/t/i-got-the-rust-compiler-to-crash-file-bug/21666) the [same ICE in simplewiki crate](https://travis-ci.org/sighol/simplewiki/builds/446856967)

**Comment by oli-obk** [comment]
I think you are encountering https://github.com/rust-lang/rust/issues/55288 which has a fix in https://github.com/rust-lang/rust/pull/55385 (making everything work again)

This issue is different, even though the symptom is the same right now. When this issue is fixed (via https://github.com/rust-lang/rust/pull/55262), the ICE will become a hard error.
